### PR TITLE
feat: apply upgrade remediation to manifests

### PR DESCRIPTION
**Pull request by kyegupov**
- [x] Ready for review
- [x] Follows CONTRIBUTING rules
- [x] Reviewed by Snyk internal team

#### What does this PR do?
Adds exported function that applies updates to the manifest files.
This has to happen in the plugin, because that's where the manifests are parsed.
Currently would only used by Registry (via pip-deps, which creates the Python environment) to prepare fix PRs.

#### Where should the reviewer start?
applyRemediationToManifests is the main entry point

### CODE REVIEW

File: test/inspect.test.js [hunk 1]
```diff
@@ -260,6 +297,52 @@ test('inspect', (t) => {
     });
 });
 
+test('inspect --only-provenance', (t) => {
+  return Promise.resolve()
```
Comment (dkontorovskyy) [inline test/inspect.test.js]: async/await maybe will be consistent ?

File: lib/manual-remediation.ts [hunk 2]
```diff
@@ -0,0 +1,16 @@
+import {applyRemediationToManifests} from '.';
```
Comment (dkontorovskyy) [inline lib/manual-remediation.ts]: Is this left over from manual testing?

File: .gitignore [hunk 3]
```diff
@@ -8,6 +8,7 @@ npm-debug.log
 *.egg-info/
 .venvs/
 dist/
+build-with-tests/
```
Comment (dkontorovskyy) [inline .gitignore]: Leftover?

File: lib/index.ts [hunk 4]
```diff
@@ -27,26 +44,161 @@ export function inspect(root, targetFile, options) {
...
+  // Other fields are of no interest
+}
+
+interface DependencyUpdates {
+  [from: string]: UpgradeRemediation;
+}
+
+interface ManifestFiles {
+  // Typically these are requirements.txt, constraints.txt and Pipfile
+  [name: string]: string; // content
```
Comment (robcresswell) [inline lib/index.ts]: Is it worth making this strict? We know what we support, so making it an arbitrary key seems unnecessary.

File: lib/index.ts [hunk 5]
```diff
@@ -27,26 +44,161 @@ export function inspect(root, targetFile, options) {
...
+            ' # not directly required, pinned by Snyk to avoid a vulnerability'
+        );
+      } else {
+        // TODO: parse constraints and replace the pre-existing one if any
+        const lines = manifests[constraintsFileName].trim().split('\n');
+        lines.push(
+          pkgName +
+            '>=' +
+            newVersion +
+            ' # pinned by Snyk to avoid a vulnerability'
```
Comment (robcresswell) [inline lib/index.ts]: We should definitely elaborate on this; the vuln id for example, but not relevant for now

File: lib/index.ts [hunk 6]
```diff
@@ -27,26 +44,161 @@ export function inspect(root, targetFile, options) {
...
+    const topLevelDep = (topLevelDeps.dependencies || {})[
+      pkgName
+    ] as DepTreeDep;
+    if (topLevelDep && topLevelDep.labels && topLevelDep.labels.provenance) {
+      const lineNubers = topLevelDep.labels.provenance
+        .split(':')[1]
+        .split('-')
+        .map((x) => parseInt(x));
+      // TODO: what if the original version spec was limited, e.g. >=1.0,<2.0 ?
+      patch[lineNubers[0] - 1] = pkgName + '>=' + newVersion;
```
Comment (robcresswell) [inline lib/index.ts]: `lineNubers` ->
`lineNumbers`

File: lib/manual-remediation.ts [hunk 7]
```diff
@@ -0,0 +1,19 @@
+/* eslint-disable no-console */
```
Comment (robcresswell) [inline lib/manual-remediation.ts]: What is this file? I'm guessing its some kind of manual test file; would prefer if this was in `test/` or `scripts/` to distinguish between this and production code, if possible

File: lib/manual-remediation.ts (hunk 2 above)
Comment (kyegupov) [reply lib/manual-remediation.ts]: Yeah, nuked.

File: .gitignore (hunk 3 above)
Comment (kyegupov) [reply .gitignore]: Nope. Let it be. Added build rules to package.json

File: test/inspect.test.js (hunk 1 above)
Comment (kyegupov) [reply test/inspect.test.js]: I'd like to upgrade the whole file to typescript but let's do that later.

File: lib/manual-remediation.ts (hunk 7 above)
Comment (kyegupov) [reply lib/manual-remediation.ts]: Yep, dropping it.

File: lib/index.ts (hunk 6 above)
Comment (kyegupov) [reply lib/index.ts]: Fixed.

File: test/inspect.test.js (hunk 1 above)
Comment (kyegupov) [reply test/inspect.test.js]: For now, extracted this test into a separate file.

File: test/remediation.test.ts [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+test('remediation with constraints.txt', async (t) => {
+  chdirWorkspaces('pip-app');
+  t.teardown(testUtils.activateVirtualenv('pip-app'));
+  const upgrades = {
+    'Django@1.6.1': { upgradeTo: 'Django@2.0.1' },
+    'transitive@1.0.0': { upgradeTo: 'transitive@1.1.1' },
+  };
+  const manifests = {
+    'requirements.txt': fs.readFileSync('requirements.txt', 'utf8'),
+    'constraints.txt': '',
```
Comment (lili2311) [inline test/remediation.test.ts]: what is `constraints.txt`?

File: test/remediation.test.ts [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+
+import plugin = require('../lib');
+import { chdirWorkspaces } from './test-utils';
+const testUtils = require('./test-utils') as any;
+
+test('remediation with constraints.txt', async (t) => {
+  chdirWorkspaces('pip-app');
+  t.teardown(testUtils.activateVirtualenv('pip-app'));
+  const upgrades = {
+    'Django@1.6.1': { upgradeTo: 'Django@2.0.1' },
```
Comment (lili2311) [inline test/remediation.test.ts]: what if you pass in `Django@1.6.1` instead of `django@1.6.1`? Can we have a test where the cases mismatch because I know this is valid in Python

File: test/remediation.test.ts [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+    'transitive@1.0.0': { upgradeTo: 'transitive@1.1.1' },
+  };
+  const manifests = {
+    'requirements.txt': fs.readFileSync('requirements.txt', 'utf8'),
+    'constraints.txt': '',
+  };
+
+  const expectedUpdatedManifests = {
```
Comment (lili2311) [inline test/remediation.test.ts]: Can the example input and output files please be a separate file in fixtures?  it is much harder to see what changes are being applied when it is all inline and what was initial input?

File: test/remediation.test.ts [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+
+  t.same(result, expectedUpdatedManifests, 'remediation as expected');
+});
+
+test('remediation without constraints.txt', async (t) => {
+  chdirWorkspaces('pip-app');
+  t.teardown(testUtils.activateVirtualenv('pip-app'));
+  const upgrades = {
+    'Django@1.6.1': { upgradeTo: 'Django@2.0.1' },
+    'transitive@1.0.0': { upgradeTo: 'transitive@1.1.1' },
```
Comment (lili2311) [inline test/remediation.test.ts]: do we support upgrading transitives? We don't for other eco systems, does this test show we do or do not support this? I can't tell

File: test/remediation.test.ts [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+    '.',
+    manifests,
+    upgrades,
+    {}
+  );
+
+  t.same(result, expectedUpdatedManifests, 'remediation as expected');
+});
+
+test('remediation without constraints.txt', async (t) => {
```
Comment (lili2311) [inline test/remediation.test.ts]: How about tests that can confirm/assert expected behaviour when unexpected things happen:
- empty upgrades object is passed?
- a computed upgrade is actually a `downgrade`
- do we know to remediate system specific packages? Like windows or mac specific ones?
- how does Python 2 / 3 play a role if at all in this remediation?

File: test/test-utils.ts [hunk 13]
```diff
@@ -1,11 +1,11 @@
-/* eslint-disable no-console */
-const fs = require('fs');
-const path = require('path');
-const process = require('process');
+/* eslint-disable no-console, @typescript-eslint/no-non-null-assertion */
+import fs = require('fs');
```
Comment (lili2311) [inline test/test-utils.ts]: please use `import and from`  style imports

File: test/test-utils.ts [hunk 14]
```diff
@@ -55,14 +55,14 @@ function activateVirtualenv(venvName) {
 function deactivateVirtualenv() {
   if (getActiveVenvName() === null) {
     console.warn('Attempted to deactivate a virtualenv when none was active.');
-    return;
+    return () => {};
```
Comment (lili2311) [inline test/test-utils.ts]: why is this needed? Does it affect what we expect to receive on the caller side?

File: test/inspect-provenance.test.ts [hunk 15]
```diff
@@ -0,0 +1,89 @@
+/* eslint-disable @typescript-eslint/no-non-null-assertion */
+import { test } from 'tap';
+import { chdirWorkspaces, activateVirtualenv } from './test-utils';
+
+import pluginImpl = require('../lib');
+
+const pipAppExpectedDependenciesOnlyProvenance = {
```
Comment (lili2311) [inline test/inspect-provenance.test.ts]: where is provenance calculated? This PR did not mention anything about it, is this a new feature in this PR?

File: lib/index.ts [hunk 16]
```diff
@@ -57,6 +57,141 @@ export async function inspect(
...
+  ) as string;
+  const constraintsFileName = Object.keys(manifests).find(
+    (fn) => path.basename(fn) === 'constraints.txt'
+  );
+
+  // Updates to requirements.txt
+  const patch: { [zeroBasedIndex: number]: string | false } = {}; // false means remove the line
+  const append: string[] = [];
+
+  for (const upgradeFrom of Object.keys(upgrades)) {
```
Comment (lili2311) [inline lib/index.ts]: this is not very readable with so many if else, can it be simplified with functions?

File: lib/index.ts [hunk 17]
```diff
@@ -57,6 +57,141 @@ export async function inspect(
...
+    const topLevelDep = (topLevelDeps.dependencies || {})[
+      pkgName
+    ] as DepTreeDep;
+    if (topLevelDep && topLevelDep.labels && topLevelDep.labels.provenance) {
+      const lineNumbers = topLevelDep.labels.provenance
+        .split(':')[1]
+        .split('-')
+        .map((x) => parseInt(x));
+      // TODO: what if the original version spec was range, e.g. >=1.0,<2.0 ?
+      patch[lineNumbers[0] - 1] = pkgName + '==' + newVersion;
```
Comment (lili2311) [inline lib/index.ts]: what does patch mean here? Because in remediation it means a fix patch, perhaps a different word is better since I assume we don't have patches yet for Python?

File: test/_setup.test.ts [hunk 18]
```diff
@@ -0,0 +1,18 @@
+import { test } from 'tap';
+import {
+  ensureVirtualenv,
+  chdirWorkspaces,
+  activateVirtualenv,
+  pipInstall,
+} from './test-utils';
+
+test('install requirements in "pip-app" venv (may take a while)', async (t) => {
```
Comment (lili2311) [inline test/_setup.test.ts]: what is this testing?

File: test/remediation.test.ts (hunk 8 above)
Comment (robcresswell) [reply test/remediation.test.ts]: `constraints` files are part of pips processing , and will be part of future remediation options. https://pip.pypa.io/en/stable/user_guide/#constraints-files

File: test/inspect-provenance.test.ts [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+      labels: { provenance: 'requirements.txt:6' },
+    },
+    msg: "testtools ok, even though it's cyclic, yay!",
+  },
+};
+
+test('inspect --only-provenance', async (t) => {
+  chdirWorkspaces('pip-app');
+  t.teardown(activateVirtualenv('pip-app'));
+  const result = await pluginImpl.inspect('.', 'requirements.txt', {
```
Comment (lili2311) [inline test/inspect-provenance.test.ts]: how about a test for what happens when this is called with Pipfile?

File: test/remediation.test.ts [hunk 20]
```diff
@@ -0,0 +1,68 @@
...
+  chdirWorkspaces('pip-app');
+  t.teardown(testUtils.activateVirtualenv('pip-app'));
+  const upgrades = {
+    'Django@1.6.1': { upgradeTo: 'Django@2.0.1' },
+    'transitive@1.0.0': { upgradeTo: 'transitive@1.1.1' },
+  };
+  const manifests = {
+    'requirements.txt': fs.readFileSync('requirements.txt', 'utf8'),
```
Comment (lili2311) [inline test/remediation.test.ts]: how about a test to show what happens when we call remediation for Pipfile? If it is not supported let's make sure it is not going to even attempt anything

File: test/remediation.test.ts (hunk 8 above)
Comment (lili2311) [reply test/remediation.test.ts]: do we support this now as well when scanning? Do we have tests to show what we support with it?

**Comment by lili2311** [comment]
Does this all support a custom file name for requirements.txt?

**Comment by kyegupov** [comment]
@lili2311 yes as long as it ends in "/requirements.txt" :) to tell it apart from "constraints.txt".

Do you think we should support arbitrary names?

**Comment by lili2311** [comment]
CLI today does support custom names. So this can be out of scope for now, but just raising this

**Comment by lili2311** [comment]
And we are about to add custom name support in SCM btw

File: test/remediation.test.ts (hunk 8 above)
Comment (kyegupov) [reply test/remediation.test.ts]: We do de-facto support them for CLI scans (because the user is free to install the deps whichever way they like).
As for the SCM support: what's the status on "project root"? Do we have a "proper" way of handling multi-file manifests?

File: test/remediation.test.ts (hunk 9 above)
Comment (kyegupov) [reply test/remediation.test.ts]: Added, seems to be working fine.

Comment (kyegupov) [reply test/remediation.test.ts]: Added a test in upper case, still works.

File: test/remediation.test.ts (hunk 10 above)
Comment (kyegupov) [reply test/remediation.test.ts]: Ok, changed.

File: test/remediation.test.ts (hunk 11 above)
Comment (kyegupov) [reply test/remediation.test.ts]: There's no reason we can't support pinning transitives in "flat" ecosystems (i.e. practically everything apart from NPM).

File: test/test-utils.ts (hunk 13 above)
Comment (kyegupov) [reply test/test-utils.ts]: Done.

File: test/test-utils.ts (hunk 14 above)
Comment (kyegupov) [reply test/test-utils.ts]: For typescript, for types to match. This function needs to return a cleanup function.

File: test/inspect-provenance.test.ts (hunk 15 above)
Comment (kyegupov) [reply test/inspect-provenance.test.ts]: It was added in the previous PR, using "--only-provenance" flag clearly visible in this test and documented in the Python code.

File: lib/index.ts (hunk 16 above)
Comment (kyegupov) [reply lib/index.ts]: I don't think functions would help here. The overhead on the repetitions and passing around parameters and having to scroll the code will not improve the readability.

File: lib/index.ts (hunk 17 above)
Comment (kyegupov) [reply lib/index.ts]: There's a comment right next to the definition of `patch`.

File: test/_setup.test.ts (hunk 18 above)
Comment (kyegupov) [reply test/_setup.test.ts]: This is setup for the tests (this is why it is in the file called `_setup.test.ts`).
This is not my idea. I have just extracted it from inspect.test.js to make sure it is run before all the tests.

File: test/inspect-provenance.test.ts (hunk 19 above)
Comment (kyegupov) [reply test/inspect-provenance.test.ts]: Added.

File: test/remediation.test.ts (hunk 20 above)
Comment (kyegupov) [reply test/remediation.test.ts]: Done.

File: test/remediation.test.ts (hunk 12 above)
Comment (kyegupov) [reply test/remediation.test.ts]: added a code to retain known "extra markers" during application of remediation and a test for one of the cases.

File: lib/index.ts [hunk 21]
```diff
@@ -57,6 +57,172 @@ export async function inspect(
...
+      'Remediation only supported for requirements.txt and constraints.txt files'
+    );
+  }
+
+  const provOptions = { ...options };
+  provOptions.args = provOptions.args || [];
+  provOptions.args.push('--only-provenance');
+
+  const topLevelDeps = (await inspect(root, targetFile, provOptions)).package;
+  applyUpgrades(manifests, upgrades, topLevelDeps);
```
[the remark this answers is not in the document]
Comment (kyegupov) [reply lib/index.ts]: I think it's probably not the plugin's job to deal with this - @robcresswell for awareness.
A way more interesting question is: what if the vuln is specified as ">1.4,<2" and we want to upgrade it to 1.6. Currently we just set it to "==1.6", which is not ideal.
